src/Servers.jl

```diff
@@ -130,278 +256,206 @@ function check_rate_limit(tcp;
+    if tcpref !== nothing
         tcpserver = tcpref[]
     elseif reuseaddr
         tcpserver = Sockets.TCPServer(; delay=false)
-        if Sys.islinux() || Sys.isapple()
+        if Sys.isunix()
             rc = ccall(:jl_tcp_reuseport, Int32, (Ptr{Cvoid},), tcpserver.handle)
-            Sockets.bind(tcpserver, host.host, host.port; reuseaddr=true)
+            Sockets.bind(tcpserver, inet.host, inet.port; reuseaddr=true)
         else
             @warn "reuseaddr=true may not be supported on this platform: $(Sys.KERNEL)"
-            Sockets.bind(tcpserver, host.host, host.port; reuseaddr=true)
+            Sockets.bind(tcpserver, inet.host, inet.port; reuseaddr=true)
         end
         Sockets.listen(tcpserver)
     else
-        tcpserver = Sockets.listen(host)
-        tcpref[] = tcpserver
+        tcpserver = Sockets.listen(inet)
     end
+    verbose && @info "Listening on: $host:$port"
 
-    if host isa Sockets.InetAddr # build debugging info
-        hostname = string(host.host)
-        hostport = string(host.port)
-    else
-        hostname = string(host)
-        hostport = ""
+    if tcpisvalid === nothing
+        tcpisvalid = ratelimit === nothing ? x->true : x->check_rate_limit(x, ratelimit)
     end
-    listenloop(f, tcpserver, sslconfig, hostname, hostport, pipeline_limit, require_ssl_verification, tcpisvalid, connectioncounter; kw...)
+
+    return listenloop(h, Server2(sslconfig, tcpserver, string(host), string(port)), tcpisvalid,
+        connectioncounter, reuse_limit, readtimeout, verbose)
 end
 
-function listenloop(f, tcpserver, sslconfig, hostname, hostport, pipeline_limit, require_ssl_verification, tcpisvalid, connectioncounter; kw...)
-    try
-        id = 0
-        while isopen(tcpserver)
-            try
-                io = accept(tcpserver)
-                if !tcpisvalid(io; kw...)
-                    @info "Accept-Reject:  $io"
-                    close(io)
-                    continue
-                end
-                io = getsslcontext(io, sslconfig)
-                let i=id, conn = Connection(hostname, hostport, pipeline_limit, 0, require_ssl_verification, io)
-                    @async try
-                        @info "Accept ($i):  $conn"
-                        connectioncounter[] += 1
-                        handle_connection(f, conn; kw...)
+"Convenience object for passing around server details"
+struct Server2{S, I}
+    ssl::S # Union{SSLConfig, Nothing}; Nothing if non-SSL
+    server::I
+    hostname::String
+    hostport::String
+end
+
+Base.isopen(s::Server2) = isopen(s.server)
+Base.close(s::Server2) = close(s.server)
+
+Sockets.accept(s::Server2{Nothing, S}) where {S} = Sockets.accept(s.server)
+Sockets.accept(s::Server2) = getsslcontext(accept(s.server), s.ssl)
+
+function getsslcontext(tcp, sslconfig)
+    ssl = MbedTLS.SSLContext()
+    MbedTLS.setup!(ssl, sslconfig)
+    MbedTLS.associate!(ssl, tcp)
+    MbedTLS.handshake!(ssl)
+    return ssl
+end
```

Comment:
> It still looks like getsslcontext has moved and the listenloop diff is corrupted by Server2 stuff; and
check_readtimeout being before handle_connection munges the handle_connection diff.

I have tried to start reviewing this diff despite the remaining reordering and whitespace changes.
The first few things I've spotted so far are:
 - The doc says `reuse_limit = nolimit`, but the code is changed to `reuse_limit::Int=1`.
 - The `pipeline_limit` option has been removed completely.
 - `require_ssl_verification` has been removed too.

The headline description what has changed in Servers.jl is "Refactoring the split between HTTP.Servers & HTTP.Handlers" and the commit description just says "simplify and cleanup".
The changes listed above don't seem to relate to the Servers/Handlers split refactor.
The fact that I'm finding changes that are not covered by the descriptions makes me feel that a more thorough review would be useful. The glitches I've found in other files support this feeling (e.g. stray `@inline` and rebase oddness) maybe other glitches are lurking in the messed up diffs.

However, I understand that you have time pressures outside HTTP.jl, and that it would take more of your time to straighten out the diffs.

If that's just not something you have time for. Let me know and I'll download the before and after, fix the reordering, and to do the comparison using local diff tools that are a bit smarter than github about whitespace-only changes etc.

Reply:
My thought on `reuse_limit` was that it might be surprising for users if they find out that the server will keep connections open indefinitely (as far as I know, this isn't a normal default in other server implementations). Better to be conservative and let the users set a higher `reuse_limit` if they have a use-case that needs it. I've actually had a similar experience on the client-side; most http client libraries don't allow automatic connection reuse.

Reply:
I think it would be better to decide how I'm going to figure out what has changed before we get into the details of discussing the few changes I've noticed so far. No doubt there will be things that seem odd at first, but after I've digested the whole change they will make more sense.
The options are:
1) I wait until you have time to revise the ordering/whitespace issues that are making the github diff review tool useless.
2) I download the branch, so some local ordering cleanup and use better diff tools
3) I do some more extensive cleanup such that the github diffs are reduced to semantic changes the push that back to the PR branch.

Reply:
Do whatever you need to do man. For me, it works best to just pull the branch locally and have the current code open in the browser; I can then see the full code in context locally and play around with things, understanding it for what it is, and using the current code in the browser for reference when I want to see how things used to be. I also tend to try and avoid stylistic comments/changes at first, because it can be frustrating for the author to get "surface-level" comments without any real review.

Reply:
I've made a few of commits on a seperate branch (based on `jq/servers`).

The first two commits don't make any semantic change but result it a github side-by-side diff that is relatively easy to follow:

https://github.com/JuliaWeb/HTTP.jl/commit/ecb6893cd882bfed0df0c2c5f87167b9b12ffbd5: move  (very nice) router example from Servers.jl to Routers.jl; and
move `handle(h, stream::Stream)` to Servers.jl  and  misc reordering / whitespace changes to reduce diff site.

https://github.com/JuliaWeb/HTTP.jl/commit/c4bddc7ffc2f4e1c621020e787adcdcfef571968: Revert stream::HTTP.Stream local variable name to http::HTTP.Stream.
This is consistent with with Streams.jl and StreamRequest.jl.
Also removes inconsistency with stream::Bool in `listen`.

If you could merge these two into your PR branch that would make it much easier for me tor review the changes.

The third commit is more speculative: https://github.com/JuliaWeb/HTTP.jl/commit/b649a3e39f8f49270b8f73ca572a00f4d9ec1b8d

This make HandlerFunction types callable.
e.g. `(h::RequestHandlerFunction)(req) = handle(h,req)`

This allows `listen` to treat the `h` handler parameter is a compeltely
generic way. i.e. functions, or Handlers or anything else that uses
the `object(args)` call pattern can be used as a request handler.

When `stream::Bool=true` the `::Stream` is passed to the handler `h` directly.
When `stream::Bool=false` the handler is wrapped a call to `handle(h, ::Stream)`.

I would really like to find a clean way to integrate Servers.jl with Handlers.jl such that Servers.jl does not depend on Handlers.jl.

Reply:
I agreed with @simonbyrne down below that the `h = let f = h; http::Stream -> handle(f, http) end` approach is pretty confusing and would require _any_ custom handler type to be callable.

I've pushed another idea:
* Move the 10 lines of code or so that define `abstract type Handler` and `RequestHandlerFunction` and `StreamHandlerFunction` into the Servers.jl module; in `listen`, pass `f` through if it's a `Handler`, or wrap it in a `XHandlerFunction` based on the `stream` argument
* We then call `handle(h, http)` in the Transaction `handle` method, which will dispatch correctly based on the provided handler argument
* I _really_ like this idea because it utilizes multiple dispatch in an elegant way (simultaneously dispatching on the type of `Handler` and whether it's handling a `Request` or `Stream` argument)
* To counter previous arguments against: 
  * Previously there was a concern about depending on the `Handlers` module to use this solution; was the issue just that `Handler`/`handle` were defined elsewhere? If so, then this is a proper solution; they live together and there are no outer module dependencies
  * If the concern was tying `listen` to use `Handlers` I have to press for a stronger argument against. In favor is the fact that this is exactly how `go` has their server setup: their `Server` type has a `Handler` field on which they call `.serveHTTP` to handle all requests. It also utilizes dispatch (idiomatic) vs. having to make all things callable (non-idiomatic).
* There were also some objections to the name `handle`, but again, I would press on that argument: it's the same terminology that [java](https://docs.oracle.com/javase/7/docs/api/javax/xml/ws/spi/http/HttpHandler.html), [rust](https://docs.rs/actix-web/0.5.6/actix_web/server/trait.HttpHandler.html), [python](https://docs.python.org/2/library/basehttpserver.html), and [go](https://golang.org/pkg/net/http/?) use, just to name a few. It's also specifically _not_ exported from the HTTP.jl package, so the usage is _always_ `HTTP.handle`, which I would posit most users would understand to mean server-side request handling.

At this point, I personally feel this is a very a) idiomatic and b) well-supported via other language implementations kind of solution; arguments against should specifically address these two advantages.